### PR TITLE
Added test file for AmpContext.js

### PR DESCRIPTION
**Pull request by bradfrizzell**
Wrote this a while back, mistakenly never merged it.

### CODE REVIEW

File: test/functional/test-amp-context.js [hunk 1]
```diff
@@ -0,0 +1,316 @@
+/**
+ * Copyright 2015 The AMP HTML Authors. All Rights Reserved.
```
Comment (alanorozco) [inline test/functional/test-amp-context.js]: 2017

File: test/functional/test-amp-context.js [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+  let win;
+  let sandbox;
+
+  beforeEach(() => {
+    sandbox = sinon.sandbox.create();
+    windowPostMessageSpy = sinon.sandbox.spy();
+    win = {
+      addEventListener: (eventType, handlerFn) => {
+        expect(eventType).to.equal('message');
+        expect(windowMessageHandler).to.not.be.ok;
```
Comment (alanorozco) [inline test/functional/test-amp-context.js]: Why do you need this check?

File: test/functional/test-amp-context.js [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+  });
+
+  afterEach(() => {
+    sandbox.restore();
+    win = undefined;
+    windowMessageHandler = undefined;
+  });
+
+  it('should add metadata to window.context using name.', () => {
+    win.name = generateStringifiedAttributes();
```
Comment (alanorozco) [inline test/functional/test-amp-context.js]: If you look at [3p-frame.js](https://github.com/ampproject/amphtml/blob/83c51ac4ddedb4e91dcfc65240d338518fcb901c/src/3p-frame.js#L89) you'll see that the object serialized to name attribute contains an "attributes" field that contains the result of this call. Confusing, I know. But this is currently not representative of the name attribute that is being passed on to the 3p frame.

File: test/functional/test-amp-context.js [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+
+    // resize denied callback should have been called
+    expect(deniedCallbackSpy.calledOnce).to.be.true;
+    expect(deniedCallbackSpy.calledWith(messagePayload));
+
+    expect(successCallbackSpy.called).to.be.false;
+  });
+});
+
+function generateStringifiedAttributes(opt_sentinel) {
```
Comment (alanorozco) [inline test/functional/test-amp-context.js]: nit: `serialized` instead of `stringified`

File: test/functional/test-amp-context.js [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+    const message = {
+      source: context.client_.hostWindow_,
+      data: messageData,
+    };
+    windowMessageHandler(message);
+
+    // window.context should have received intersection observer postMessage
+    // back, and should have called the callback function
+    expect(callbackSpy.calledOnce).to.be.true;
+    expect(callbackSpy.calledWith(messagePayload));
```
Comment (alanorozco) [inline test/functional/test-amp-context.js]: Side note.

There's a small problem with this that will be fixed once #8469 is merged. The callback for `observeIntersection` in the current window.context API will receive an array of changes, not a change. This should not be a problem for this PR but just letting you know of the current API incompatibility.

File: test/functional/test-amp-context.js [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+
+function generateAttributes(opt_sentinel) {
+  const attributes = {};
+  const sentinel = opt_sentinel || '1-291921';
+  attributes._context = {
+    location: 'foo.com',
+    canonicalUrl: 'foo.com',
+    pageViewId: '1',
+    sentinel,
+    startTime: '0',
```
Comment (alanorozco) [inline test/functional/test-amp-context.js]: I think this should be a number, not a string.

File: test/functional/test-amp-context.js (hunk 3 above)
Comment (bradfrizzell) [inline test/functional/test-amp-context.js]: o.O
Good catch. That is actually a bigger problem than just for this PR, the AmpContext library is parsing improperly overall. I have no idea how I missed this. Will submit a separate PR to fix this ASAP (and will also address the local issues in this PR).

File: test/functional/test-amp-context.js (hunk 2 above)
Comment (bradfrizzell) [inline test/functional/test-amp-context.js]: useful as a sanity check that we only try to test one message per test. If you try to register a second event listener in one test, as it is right now, it will error instead of just letting you overwrite the initial message handler.

File: test/functional/test-amp-context.js (hunk 6 above)
Comment (bradfrizzell) [inline test/functional/test-amp-context.js]: you're right, done.

File: test/functional/test-amp-context.js (hunk 4 above)
Comment (bradfrizzell) [inline test/functional/test-amp-context.js]: done

File: test/functional/test-amp-context.js (hunk 5 above)
Comment (bradfrizzell) [inline test/functional/test-amp-context.js]: noted

File: test/functional/test-amp-context.js (hunk 1 above)
Comment (bradfrizzell) [inline test/functional/test-amp-context.js]: done

**Comment by bradfrizzell** [comment]
ping on merging if there are no further issues
